### PR TITLE
add fetch-failed-logs script

### PR DESCRIPTION
**Pull request by dawidd6**
This script fetches latest build failures logs from GitHub Actions API for given formula.

Supports `repository_dispatch` (`--dispatched` option) and `pull_request` events.

Example:

```
➜  ~  brew fetch-failed-logs imapsync --dispatched
==> Workflow run id: 67726873
==> brew test imapsync --verbose
Testing imapsync
==> /home/linuxbrew/.linuxbrew/Cellar/imapsync/1.977/bin/imapsync --dry
Can't locate Socket6.pm in @INC (you may need to install the Socket6 module) (@INC contains: /home/linuxbrew/.linuxbrew/Cellar/imapsync/1.977/libexec/lib/perl5 /etc/perl /usr/local/lib/x86_64-linux-gnu/perl/5.22.1 /usr/local/share/perl/5.22.1 /usr/lib/x86_64-linux-gnu/perl5/5.22 /usr/share/perl5 /usr/lib/x86_64-linux-gnu/perl/5.22 /usr/share/perl/5.22 /usr/local/lib/site_perl /usr/lib/x86_64-linux-gnu/perl-base .) at /home/linuxbrew/.linuxbrew/Cellar/imapsync/1.977/libexec/lib/perl5/IO/Socket/INET6.pm line 40.
BEGIN failed--compilation aborted at /home/linuxbrew/.linuxbrew/Cellar/imapsync/1.977/libexec/lib/perl5/IO/Socket/INET6.pm line 40.
Compilation failed in require at /home/linuxbrew/.linuxbrew/Cellar/imapsync/1.977/libexec/bin/imapsync line 865.
BEGIN failed--compilation aborted at /home/linuxbrew/.linuxbrew/Cellar/imapsync/1.977/libexec/bin/imapsync line 865.
Error: imapsync: failed
An exception occurred within a child process:
Test::Unit::AssertionFailedError: </1\.977/> was expected to be =~
<"">.
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3/lib/ruby/gems/2.6.0/gems/test-unit-3.2.9/lib/test/unit/assertions.rb:55:in `block in assert_block'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3/lib/ruby/gems/2.6.0/gems/test-unit-3.2.9/lib/test/unit/assertions.rb:1631:in `_wrap_assertion'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3/lib/ruby/gems/2.6.0/gems/test-unit-3.2.9/lib/test/unit/assertions.rb:53:in `assert_block'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3/lib/ruby/gems/2.6.0/gems/test-unit-3.2.9/lib/test/unit/assertions.rb:543:in `block in assert_match'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3/lib/ruby/gems/2.6.0/gems/test-unit-3.2.9/lib/test/unit/assertions.rb:1636:in `_wrap_assertion'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3/lib/ruby/gems/2.6.0/gems/test-unit-3.2.9/lib/test/unit/assertions.rb:533:in `assert_match'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/imapsync.rb:164:in `block in <class:Imapsync>'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:1777:in `block (3 levels) in run_test'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/utils.rb:478:in `with_env'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:1776:in `block (2 levels) in run_test'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:867:in `with_logging'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:1775:in `block in run_test'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:2020:in `block in mktemp'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/mktemp.rb:57:in `block in run'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/mktemp.rb:57:in `chdir'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/mktemp.rb:57:in `run'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:2019:in `mktemp'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:1769:in `run_test'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/test.rb:35:in `block in <main>'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3/lib/ruby/2.6.0/timeout.rb:93:in `block in timeout'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3/lib/ruby/2.6.0/timeout.rb:33:in `block in catch'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3/lib/ruby/2.6.0/timeout.rb:33:in `catch'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3/lib/ruby/2.6.0/timeout.rb:33:in `catch'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3/lib/ruby/2.6.0/timeout.rb:108:in `timeout'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/test.rb:34:in `<main>'
==> FAILED
```

### CODE REVIEW

**Comment by dawidd6** [comment]
Depends on: https://github.com/Homebrew/brew/pull/7252

File: cmd/fetch-failed-logs.rb [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        commit_files.find do |file|
+          file[%r{Formula/(.+)\.rb}, 1] == formula
+        end
+        # Not supported event
+      end
+    end
+
+    unless workflow_run
+      opoo "No workflow run matching the criteria was found"
+      return
```
Comment (maxim-belkin) [inline cmd/fetch-failed-logs.rb:91]: `odie` ?

Comment (maxim-belkin) [reply cmd/fetch-failed-logs.rb:91]: `opoo` + `return` is fine too if you don't want to return an error when there is no matching workflow run.

**Comment by iMichka** [comment]
I ran this build twice: https://github.com/Homebrew/linuxbrew-core/pull/19986

After the second run, the logs can not be fetched anymore.

**Comment by dawidd6** [comment]
> I ran this build twice: [Homebrew/linuxbrew-core#19986](https://github.com/Homebrew/linuxbrew-core/pull/19986)
> 
> After the second run, the logs can not be fetched anymore.

Fixed.

File: cmd/fetch-failed-logs.rb (hunk 1 above)
Comment (dawidd6) [reply cmd/fetch-failed-logs.rb]: Suggestion applied, thanks!

**Comment by iMichka** [comment]
Thanks @dawidd6 !

@jonchang would this prove useful on mac too, especially when we have gotten rid of jenkins, to ease log reading?

**Comment by jonchang** [comment]
Yeah I think this would be super useful. I’ll think about how to adapt it so it works on both Linux and macOS...

**Pull request merged**